### PR TITLE
Add ability to create on_commit_msg policies per branches

### PR DESCRIPTION
**Pull request by charbonnierg**
This PR is an attempt to support the ability to define different tagging policies according to the current git branch.

### CODE REVIEW

**Comment by charbonnierg** [comment]
Todos:

- [ ] Complete usage example

File: src/releaser/hexagon/entities/strategy/policy.py [hunk 1]
```diff
@@ -93,18 +93,47 @@ def parse_dict(cls, data: dict[str, Any]) -> "CommitMsgMatchPolicy":
...
+@dataclass(frozen=True)
+class Rule:
+    """A rule for a release strategy."""
+
+    branch: list[str] = field(default_factory=list)
+    """The branches to match for the HEAD commit message."""
+
+    commit_msg: list[CommitMsgMatchPolicy] = field(default_factory=list)
+    """The tags to match for the HEAD commit message.
+    These are global policies that apply to all applications.
```
Comment (charbonnierg) [inline src/releaser/hexagon/entities/strategy/policy.py:121]: Obsolete comment

File: src/releaser/hexagon/ports/git_reader.py [hunk 2]
```diff
@@ -33,3 +39,9 @@ def read_last_commit_message(self, depth: int, filter: str | None) -> str | None
                     return commit_msg
             return None
         return history[0]
+
+    def current_reference_matches(self, ref: list[str]) -> bool:
+        """Check if the current reference matches the given ref."""
+        if not ref:
+            return True
+        return self.read_current_branch() in ref
```
Comment (charbonnierg) [inline src/releaser/hexagon/ports/git_reader.py:47]: This is a minimal effort: we don't support `*` character in branch name. For me it's ok

**Comment by pisua** [comment]
here the new syntax that take in account the rule with branch an d commit_msg. 

e.g with next in  toml syntax. 

````toml
[tool.quara.releaser]
on = [{ branch = ["next"], commit_msg = [
    { match = "*", tags = [
        { value = "next" },
        { type = "git_commit_sha", size = 7 },
    ] },
    { match = [
        "chore\\(release\\): bump to version [0-9]*.[0-9]*.[0-9]*\\-rc\\.[0-9]*\\s",
    ], tags = [
        { value = "edge" },
    ] },
    { match = [
        "chore\\(release\\): bump to version [0-9]*.[0-9]*.[0-9]*",
        "chore\\(release\\): merge from stable branch",
    ], tags = [
        { type = "version" },
    ] },
    { match = [
        "chore\\(release\\): bump to version [0-9]*.[0-9]*.[0-9]*\\s",
    ], tags = [
        { value = "latest" },
    ] },
] }]
````

**Comment by charbonnierg** [comment]
Would it be better to use a dict instead ?

```toml
[tool.quara.releaser.on.continuous-delivery]
branch = ["next"]
commit_msg = [
    { match = "*", tags = [
        { value = "next" },
        { type = "git_commit_sha", size = 7 },
    ] },
    { match = [
        "chore\\(release\\): bump to version [0-9]*.[0-9]*.[0-9]*\\-rc\\.[0-9]*\\s",
    ], tags = [
        { value = "edge" },
    ] },
    { match = [
        "chore\\(release\\): bump to version [0-9]*.[0-9]*.[0-9]*",
        "chore\\(release\\): merge from stable branch",
    ], tags = [
        { type = "version" },
    ] },
    { match = [
        "chore\\(release\\): bump to version [0-9]*.[0-9]*.[0-9]*\\s",
    ], tags = [
        { value = "latest" },
    ]}]
```

**Comment by charbonnierg** [comment]
I'm worried that `continous-delivery` has literally no impact aside from allowing to write better TOML...

**Comment by charbonnierg** [comment]
In JSON, it would become:

```json
{
  "on": {
    "continuous-delivery": {
      "branch": [
        "next"
      ],
      "commit_msg": [
        {
          "match": "*",
          "tags": [
            {
              "value": "next"
            },
            {
              "type": "git_commit_sha",
              "size": 7
            }
          ]
        }
      ]
    },
    "something-else": {
      "commit_msg": [
        {
          "match": "*",
          "tags": [
            {
              "type": "git_commit_sha",
              "size": 7
            }
          ]
        }
      ]
    }
  }
}
```

BTW do we expect a single `on` rule to be applied, or do we expect all rules that match to be applied ?

**Comment by charbonnierg** [comment]
```json
{
  "on": {
    "continuous-delivery": {
      "branch": [
        "next"
      ],
      "commit_msg": [
        {
          "match": "*",
          "tags": [
            {
              "value": "next"
            },
            {
              "type": "git_commit_sha",
              "size": 7
            }
          ]
        }
      ]
    },
    "something-else": {
      "exclude_branch": ["next"],
      "commit_msg": [
        {
          "match": "*",
          "tags": [
            {
              "type": "git_commit_sha",
              "size": 7
            }
          ]
        }
      ]
    }
  }
}
```

**Comment by charbonnierg** [comment]
Should we also add an `exclude_branch`  to allow non-overlaping rules ?

**Comment by charbonnierg** [comment]
Proposed rules:

- Either `branch` or `exclude_branch` must be present

- If both `branch`  and `exclude_branch` are specified, we need to decide whether we want to:
  - raise an error
  - or take the included banches minus the excluded branches. That's not really difficult so we might want to support that.

In the end, all `on` entries that matches the current git state (e.g., both expected git branch and commit messages match), then they are all applied.

**Comment by charbonnierg** [comment]
Let's slow down a minute and think about the use cases:

- We want to be able to match `next` branch only
- We want to be able to match all branches except `next`

I don't see any other needs at the moment.

**Comment by charbonnierg** [comment]
Even there it's debatable, maybe what we need is simply:

- match `next` only
- match all branches

**Comment by charbonnierg** [comment]
To conclude, we could use:

```json
{
  "on": {
    "continuous-delivery": {
      "branch": [
        "next"
      ],
      "commit_msg": [
        {
          "match": "*",
          "tags": [
            {
              "value": "next"
            }
          ]
        }
      ]
    },
    "something-else": {
      "commit_msg": [
        {
          "match": "*",
          "tags": [
            {
              "type": "git_commit_sha",
              "size": 7
            }
          ]
        }
      ]
    }
  }
}
```

`git_commit_sha` is applied to all branches. And there are some specific tags for `next` branch only

**Comment by pisua** [comment]
The equivalent in toml will be : 

````
[tool.quara.releaser.on.continuous-delivery]
branch = ["next"]
commit_msg = [
    { match = "*", tags = [
        { value = "next" }
    ] },
    { match = [
        "chore\\(release\\): bump to version [0-9]*.[0-9]*.[0-9]*\\-rc\\.[0-9]*\\s",
    ], tags = [
        { value = "edge" }
    ] },
    { match = [
        "chore\\(release\\): bump to version [0-9]*.[0-9]*.[0-9]*",
        "chore\\(release\\): merge from stable branch",
    ], tags = [
        { type = "version" }
    ] },
    { match = [
        "chore\\(release\\): bump to version [0-9]*.[0-9]*.[0-9]*\\s",
    ], tags = [
        { value = "latest" }
    ] },
]

[tool.quara.releaser.on.all]
commit_msg = [
    { match = "*", tags = [
        { type = "git_commit_sha", size = 7 }
    ] }
]
````

**Pull request merged**